onnxruntime/core/providers/coreml/builders/impl/pad_op_builder.cc

```diff
@@ -0,0 +1,244 @@
+  const auto& input_defs = node.InputDefs();
+  const auto& initializers = input_params.graph_viewer.GetAllInitializedTensors();
+
+  std::vector<int64_t> input_shape;
+  if (!GetShape(*input_defs[0], input_shape, logger))
+    return false;
+
+  if (input_shape.empty() || input_shape.size() < 2) {
+    LOGS(logger, VERBOSE) << "Pad requires input shape to be at least 2d, input is "
+                          << input_shape.size() << "d shape";
+    return false;
+  }
+
+  if (std::find(input_shape.begin(), input_shape.end(), int64_t{0}) != input_shape.end()) {
+    LOGS(logger, VERBOSE) << "Input with a 0 for a dimension is not supported";
+    return false;
+  }
+
+  {
+    NodeAttrHelper helper(node);
+    const auto mode = helper.Get("mode", "constant");
+    if (mode != "constant") {
+      LOGS(logger, VERBOSE) << "Only `constant` mode Pad is currently supported for now, mode: " << mode;
+      return false;
+    }
+
+    if (input_defs.size() < 3) {
+      LOGS(logger, VERBOSE) << "`constant_value` input is required for constant mode Pad op.";
+      return false;
+    }
+  }
+
+  // only support if `pads` input is known and does not contain negative values and only applies padding values
+  // for last two dimensions.
+  {
+    const auto pads_initializer_it = initializers.find(input_defs[1]->Name());
+    if (pads_initializer_it == initializers.end()) {
+      LOGS(logger, VERBOSE) << "pads must be a constant initializer.";
+      return false;
+    }
+
+    const ONNX_NAMESPACE::TensorProto& pads_initializer = *pads_initializer_it->second;
+    Initializer unpacked_tensor(pads_initializer);
+
+    auto pads_tensor_data = unpacked_tensor.DataAsSpan<int64_t>();
+    for (int64_t i = 0; i < unpacked_tensor.size(); i++) {
+      if (pads_tensor_data[i] < 0) {
+        LOGS(logger, VERBOSE) << "Negative pad value is not supported: pads["
+                              << i << "] = " << pads_tensor_data[i];
+        return false;
+      }
+    }
+
+    std::variant<InlinedVector<int64_t>, gsl::span<const int64_t>> axes_tensor_data;
+    // Check that only supports padding on last two dimensions - [H,W].
+    // CoreML PaddinglayerParams: https://apple.github.io/coremltools/mlmodel/Format/NeuralNetwork.html#paddinglayerparams
+    const auto input_rank = input_shape.size();
+    int64_t num_axes = input_rank;
+    if (input_defs.size() > 3) {
+      // optional axes input is provided
+      const auto axes_initializer_it = initializers.find(input_defs[3]->Name());
+      if (axes_initializer_it == initializers.end()) {
+        LOGS(logger, VERBOSE) << "If provided, axes must be a constant initializer.";
+        return false;
+      }
+      const ONNX_NAMESPACE::TensorProto& axes_tensor = *axes_initializer_it->second;
+      Initializer axes_initializer(axes_tensor);
+      auto axes_span = axes_initializer.DataAsSpan<int64_t>();
+      axes_tensor_data.emplace<1>(axes_span);
```

Comment:
span doesn't copy, so do we still have a scope issue here? axes_span is over data in axes_initializer which goes out of scope leaving the value in axes_tensor_data pointing to invalid memory. I think you need to create an InlinedVector from the data in axes_initializer so there's a copy made. 

Given that should we just use InlinedVector for axes_tensor_data in both cases instead of having a std::variant?

Reply:
thanks for pointing out this. okay I created an InlinedVector for holding axes tensor data, and reverted back some changes. hopefully this way can keep things slightly simpler.